File: py/fetch_data.py
"""
抓取双色球历史开奖数据 -> data/sample_history.csv

数据源：500 彩票网公开历史接口（无需登录、无需 cookie）
用法：
    python fetch_data.py            # 默认抓取最近 200 期
    python fetch_data.py 500        # 抓取最近 500 期

输出 CSV 列：issue,date,red,blue
    red 为 "01,05,06,10,12,23" 形式（逗号分隔、两位补零）
"""

from __future__ import annotations

import csv
import os
import sys
import urllib.request
from html.parser import HTMLParser

BASE_URL = "https://datachart.500.com/ssq/history/newinc/history.php"
OUT_PATH = os.path.join(os.path.dirname(os.path.abspath(__file__)), "data", "sample_history.csv")

# HTML 表格行解析：收集每个 <tr> 内的 <td> 文本
class _TableParser(HTMLParser):
    def __init__(self) -> None:
        super().__init__()
        self.rows: list[list[str]] = []
        self._row: list[str] = []
        self._in_td = False
        self._buf = ""

    def handle_starttag(self, tag, attrs):
        if tag == "tr":
            self._row = []
        elif tag == "td":
            self._in_td = True
            self._buf = ""

    def handle_endtag(self, tag):
        if tag == "td":
            self._row.append(self._buf.strip())
            self._in_td = False
        elif tag == "tr":
            if self._row:
                self.rows.append(self._row)

    def handle_data(self, data):
        if self._in_td:
            self._buf += data


def _to_ints(cells: list[str]) -> list[int]:
    out = []
    for c in cells:
        c = c.strip()
        if c.isdigit():
            out.append(int(c))
    return out


def parse_rows(html: str) -> list[dict]:
    """
    从 500 历史页 HTML 提取 (issue, date, reds[6], blue)。

    ⚠️ 必须按【列位置】解析，不能按数值范围猜。
    500 历史表每行 16 列，结构固定：
      [0]  期号        26087
      [1]-[6] 红球     04 06 10 18 23 31
      [7]  蓝球        11
      [8]  快乐星期天
      [9]  奖池奖金
      [10] 一等奖注数  [11] 一等奖奖金
      [12] 二等奖注数  [13] 二等奖奖金
      [14] 总投注额
      [15] 开奖日期    2026-07-30

    历史教训：早期版本用「第一个 1-16 之间的数字」当蓝球，
    结果永远取到了第 1 个红球（红球也可能 <=16），
    导致整份数据的蓝球列 = 最小红球，走势分析与回测全部失真。
    """
    p = _TableParser()
    p.feed(html)
    results = []
    for row in p.rows:
        if len(row) < 8:
            continue
        issue = row[0].strip()
        if not (issue.isdigit() and len(issue) >= 5):
            continue

        # --- 严格按位置取号并校验 ---
        red_cells = [c.strip() for c in row[1:7]]
        blue_cell = row[7].strip()
        if not all(c.isdigit() for c in red_cells) or not blue_cell.isdigit():
            continue
        reds = [int(c) for c in red_cells]
        blue = int(blue_cell)

        # 规则校验：红球 6 个互不重复且在 1-33；蓝球在 1-16
        if len(set(reds)) != 6:
            continue
        if not all(1 <= n <= 33 for n in reds):
            continue
        if not (1 <= blue <= 16):
            continue

        date = next((c.strip() for c in row if c.count("-") == 2 and c.strip()[:4].isdigit()), "")
        results.append(
            {
                "issue": issue,
                "date": date,
                "red": ",".join(f"{n:02d}" for n in sorted(reds)),
                "blue": f"{blue:02d}",
            }
        )
    return results


def sanity_check(rows: list[dict]) -> list[str]:
    """
    对抓取结果做数据质量体检，返回告警列表（空列表 = 通过）。
    重点防御「列错位」这类静默数据污染。
    """
    warns: list[str] = []
    if not rows:
        return ["未解析到任何数据"]

    from collections import Counter

    n = len(rows)
    blues = [int(r["blue"]) for r in rows]
    bc = Counter(blues)

    # 1) 蓝球必须覆盖较多号码；若集中在少数几个，说明列取错了
    if n >= 100 and len(bc) < 14:
        warns.append(f"蓝球只出现 {len(bc)}/16 种取值，疑似列错位")

    # 2) 蓝球应近似均匀（期望 n/16）。最大频次远超期望即异常
    exp = n / 16
    if n >= 100 and max(bc.values()) > exp * 2.2:
        top = bc.most_common(1)[0]
        warns.append(f"蓝球 {top[0]:02d} 出现 {top[1]} 次，远超期望 {exp:.1f} 次，疑似列错位")

    # 3) 蓝球不应恒等于最小红球（这正是历史 bug 的特征）
    same = sum(1 for r in rows if int(r["blue"]) == int(r["red"].split(",")[0]))
    if same > n * 0.3:
        warns.append(f"{same}/{n} 期的蓝球等于最小红球，几乎可以断定解析错列")

    # 4) 红球整体频率应接近 n*6/33
    from collections import Counter as C2

    rc = C2()
    for r in rows:
        rc.update(int(x) for x in r["red"].split(","))
    if n >= 100:
        exp_r = n * 6 / 33
        if max(rc.values()) > exp_r * 1.8 or len(rc) < 33:
            warns.append("红球频率分布异常，可能存在数据污染")

    return warns


def fetch(limit: int = 200) -> list[dict]:
    if not (5 <= limit <= 99999):
        raise ValueError(f"历史期数需在 5–99999 期之间（当前请求 {limit}）")
    # 500 接口：start/end 为期号范围，limit 限制条数
    url = f"{BASE_URL}?start=23001&end=26999&limit={limit}"
    req = urllib.request.Request(url, headers={"User-Agent": "Mozilla/5.0"})
    with urllib.request.urlopen(req, timeout=20) as resp:
        html = resp.read().decode("gb2312", errors="ignore")
    rows = parse_rows(html)
    # 500 接口默认按【期号降序】返回（最新在前）。
    # 去重（同一期可能因分页边界重复），并保留最新 limit 期。
    seen = set()
    dedup = []
    for r in rows:
        if r["issue"] in seen:
            continue
        seen.add(r["issue"])
        dedup.append(r)
    return dedup[:limit] if limit else dedup


def main() -> int:
    limit = int(sys.argv[1]) if len(sys.argv) > 1 else 200
    print(f"正在抓取最近 {limit} 期双色球历史数据 ...")
    try:
        rows = fetch(limit)
    except Exception as e:  # noqa: BLE001
        print(f"✗ 抓取失败：{e}")
        print("  可能原因：当前环境无外网访问，或数据源临时不可用。")
        print("  可手动从官网导出历史开奖数据，按 data/sample_history.csv 格式保存后使用。")
        return 1

    if not rows:
        print("✗ 未解析到任何数据，请检查数据源结构是否变化。")
        return 1

    warns = sanity_check(rows)
    if warns:
        print("✗ 数据质量体检未通过，已中止写入（避免污染历史库）：")
        for w in warns:
            print(f"    - {w}")
        return 1
    print("✓ 数据质量体检通过（蓝球分布均匀、红蓝列未错位）")

    os.makedirs(os.path.dirname(OUT_PATH), exist_ok=True)
    with open(OUT_PATH, "w", newline="", encoding="utf-8-sig") as f:
        w = csv.DictWriter(f, fieldnames=["issue", "date", "red", "blue"])
        w.writeheader()
        w.writerows(rows)

    print(f"✓ 已写入 {len(rows)} 期到 {OUT_PATH}")
    print("  最近 3 期预览：")
    for r in rows[:3]:
        print(f"    {r['issue']}  {r['date']}  红:{r['red']}  蓝:{r['blue']}")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: functions/api/fetch_data.js
// 抓取双色球历史开奖数据（对应 py/fetch_data.py 的 JS 移植）
//
// 数据源：500 彩票网公开历史接口（无需登录、无需 cookie）。
// 关键防御：必须【按列位置】解析，不能按数值范围猜 —— 历史上吃过「蓝球列错位」
// 的大亏（详见 py/fetch_data.py 顶部说明），一旦写错会静默污染整个历史库。
//
// 编码说明：500 接口返回 gb2312。在 Cloudflare Workers / Node 中我们用 utf-8 解码，
// 因为开奖数据里我们真正需要的 期号 / 红球 / 蓝球 / 日期 全是 ASCII（数字与短横线），
// 在 gb2312 与 utf-8 下字节完全一致，所以中文部分虽会变成乱码，但解析结果不受影响。

const BASE_URL = "https://datachart.500.com/ssq/history/newinc/history.php";

// HTML 表格行解析：收集每个 <tr> 内的 <td> 文本
function extractRows(html) {
  // ⚠️ 必须先剥离 HTML 注释：500 彩票网在近期的历史表每行开头塞了
  // `<!--<td>2</td>-->` 之类的注释残留，正则 <td> 会把注释里的 <td> 也当成
  // 一格，导致整行右移一列、期号/蓝球错位 → 解析出 0 行。
  // 这与 py 版（用 HTMLParser，天然忽略注释）行为保持一致。
  html = html.replace(/<!--[\s\S]*?-->/g, "");
  const rows = [];
  const trRe = /<tr[^>]*>([\s\S]*?)<\/tr>/gi;
  let trM;
  while ((trM = trRe.exec(html))) {
    const tds = [];
    const tdRe = /<td[^>]*>([\s\S]*?)<\/td>/gi;
    let tdM;
    while ((tdM = tdRe.exec(trM[1]))) {
      const text = tdM[1].replace(/<[^>]+>/g, "").trim();
      tds.push(text);
    }
    if (tds.length) rows.push(tds);
  }
  return rows;
}

// 从 500 历史页 HTML 提取 (issue, date, reds[6], blue)
// ⚠️ 严格按【列位置】解析：见 py/fetch_data.py 的列结构注释
export function parseRows(html) {
  const results = [];
  for (const row of extractRows(html)) {
    if (row.length < 8) continue;
    const issue = row[0].trim();
    if (!(issue.length >= 5 && /^\d+$/.test(issue))) continue;

    const redCells = row.slice(1, 7).map((c) => c.trim());
    const blueCell = row[7].trim();
    if (!redCells.every((c) => /^\d+$/.test(c)) || !/^\d+$/.test(blueCell)) continue;
    const reds = redCells.map(Number);
    const blue = Number(blueCell);

    // 规则校验：红球 6 个互不重复且在 1-33；蓝球在 1-16
    if (new Set(reds).size !== 6) continue;
    if (!reds.every((n) => n >= 1 && n <= 33)) continue;
    if (!(blue >= 1 && blue <= 16)) continue;

    const date =
      row.find(
        (c) => (c.match(/-/g) || []).length === 2 && /^\d{4}/.test(c.trim())
      ) || "";
    results.push({
      issue,
      date,
      red: reds
        .slice()
        .sort((a, b) => a - b)
        .map((n) => String(n).padStart(2, "0"))
        .join(","),
      blue: String(blue).padStart(2, "0"),
    });
  }
  return results;
}

// 数据质量体检，返回告警列表（空 = 通过）。重点防御「列错位」这类静默污染。
export function sanityCheck(rows) {
  const warns = [];
  if (!rows.length) return ["未解析到任何数据"];

  const n = rows.length;
  const bc = {};
  for (const r of rows) {
    const b = parseInt(r.blue, 10);
    bc[b] = (bc[b] || 0) + 1;
  }

  // 1) 蓝球必须覆盖较多号码
  const blueKinds = Object.keys(bc).length;
  if (n >= 100 && blueKinds < 14)
    warns.push(`蓝球只出现 ${blueKinds}/16 种取值，疑似列错位`);

  // 2) 蓝球应近似均匀（期望 n/16）
  const exp = n / 16;
  let topB = 0, topBv = 0;
  for (const k in bc) if (bc[k] > topBv) { topBv = bc[k]; topB = Number(k); }
  if (n >= 100 && topBv > exp * 2.2)
    warns.push(
      `蓝球 ${String(topB).padStart(2, "0")} 出现 ${topBv} 次，远超期望 ${exp.toFixed(1)} 次，疑似列错位`
    );

  // 3) 蓝球不应恒等于最小红球（历史 bug 的特征）
  const same = rows.filter(
    (r) => parseInt(r.blue, 10) === parseInt(r.red.split(",")[0], 10)
  ).length;
  if (same > n * 0.3)
    warns.push(`${same}/${n} 期的蓝球等于最小红球，几乎可以断定解析错列`);

  // 4) 红球整体频率应接近 n*6/33
  const rc = {};
  for (const r of rows)
    for (const x of r.red.split(",")) rc[x] = (rc[x] || 0) + 1;
  if (n >= 100) {
    const expR = (n * 6) / 33;
    let maxRv = 0;
    for (const k in rc) if (rc[k] > maxRv) maxRv = rc[k];
    if (maxRv > expR * 1.8 || Object.keys(rc).length < 33)
      warns.push("红球频率分布异常，可能存在数据污染");
  }

  return warns;
}

// 校验历史期数：最小 5 期，最大 99999 期（与 py 版一致）。非法则抛错。
// 严格模式：必须是正整数（拒绝小数/字母/负数/0），不再用 parseInt 宽松解析。
export function validateLimit(n) {
  if (typeof n !== "string") n = String(n ?? "");
  n = n.trim();
  if (!/^\d+$/.test(n)) throw new Error("历史期数必须是 5–99999 之间的正整数");
  const v = parseInt(n, 10);
  if (v < 5 || v > 99999) throw new Error("历史期数需在 5–99999 期之间");
  return v;
}

// 抓取最近 limit 期（默认 500）。500 接口按期号降序返回（最新在前）。
export async function fetchLatest(limit = 500) {
  const count = validateLimit(limit);
  const url = `${BASE_URL}?start=23001&end=26999&limit=${count}`;
  const res = await fetch(url, {
    headers: { "User-Agent": "Mozilla/5.0" },
  });
  if (!res.ok) throw new Error("抓取失败：HTTP " + res.status);
  const buf = await res.arrayBuffer();
  const html = new TextDecoder("utf-8").decode(buf);
  const rows = parseRows(html);

  // 去重（分页边界可能重复），保留最新 limit 期
  const seen = new Set();
  const dedup = [];
  for (const r of rows) {
    if (seen.has(r.issue)) continue;
    seen.add(r.issue);
    dedup.push(r);
  }
  return limit ? dedup.slice(0, limit) : dedup;
}

export function toCSV(rows) {
  // red 字段内嵌逗号，必须用引号包裹，与静态样本（py 以 csv 模块写出）格式一致，
  // 这样 engine.parseCSV 对「静态样本」与「KV 刷新后」两种来源都能正确解析。
  const lines = ["issue,date,red,blue"];
  for (const r of rows)
    lines.push(`${r.issue},${r.date},"${r.red}",${r.blue}`);
  return lines.join("\n");
}
